design/BufferOperations.md

```diff
@@ -0,0 +1,148 @@
+ - `success` is called with the content of the memory as an argument.
+
+If `success` hasn't been called when the WebGPUMappedMemory gets invalidated (meaning the object is still in the pending state), `error` is called instead. When `WebGPUMappedMemory` goes from the available state to the invalidated state, the `ArrayBuffer` for its content gets neutered. The return value of `then` acts like the return value of `Promise.then`.
+
+## Immediate data upload
+
+Buffer mapping is the path with the least number of copies but it is often useful to upload data to a buffer *right now*, if only for debugging.
+A `WebGPUBuffer` operation is provided that takes an ArrayBuffer and copies its content at an offset in the buffer.
+
+```
+partial interface WebGPUBuffer {
+    void setSubData(ArrayBuffer data, u32 offset);
+}
+```
+
+This operation acts as if it was done after all previous "device-level" commands and before all subsequent "device-level" commands.
+"Device level" commands are all commands not buffered in a `WebGPUCommandBuffer`, and include `WebGPUQueue.submit`.
+The content of `data` is only read during the call and can be modified by the application afterwards.
+The following must be true or a validation error occurs:
+
+ - The buffer must have been created with the `WebGPUBufferUsage.TRANSFER_DST` usage flag.
+ - `offset + data.length` must not overflow and be at most the size of the buffer.
+ - Depending on the design of memory barriers, the buffer must be, or allowed to be in the `WebGPUBufferUsage.TRANSFER_DST` usage.
+   - In particular the buffer must not be currently mapped.
+
+## Unused designs
+
+### Persistently mapped buffer
+
+Persistently mapped buffer are when the result of mapping the buffer can be kept by the application while the buffer is in use by the GPU.
+We didn't find a way to have persistently mapped buffers and at the same time keep things data race free between the CPU and GPU.
+Being data race free would be possible if ArrayBuffer could be unneutered but this is not the case.
+
+### Promise<ArrayBuffer> readback();
+
+This didn't have a pollable interface and forced an extra buffer-to-buffer copy to occur if the GPU execution could be resumed immediately.
+
+### NXT's MapReadAsync(callback);
+
+Not a pollable interface.
+
+## Issues
+
+### GC discoverability
+
+It isn't clear yet what happens when a buffer gets garbage collected while it is mapped.
+The simple answer is that the `WebGPUMappedMemory` objects get invalidated but that would allow the application to discover when the GC runs.
+
+### GC pressure
+
+The `WebGPUMappedMemory` design makes each mapped region create two garbage collected objects. This could lead to some GC pressure.
+
+### Side effects between mapped memory regions
+
+What happens when `WebGPUMappedMemory` object's region in the buffer overlap?
+Are write from one visible from the other?
+If they are, maybe `WebGPUMappedMemory.getPointer` should return an `ArrayBufferView` instead.
+
```

Comment:
Thanks I didn't know that. This might mean that we need to require mapped ranges to be disjoint but that's a pretty big restriction. @jdashg's idea of more explicit "client buffers" become more attractive.